src/pools/D3MAaveDaiPool.sol

```diff
@@ -116,28 +118,30 @@ contract D3MAaveDaiPool is D3MPoolBase {
     }
 
     // --- Collect any rewards ---
-    function collect(address[] memory assets, uint256 amount) external auth returns (uint256 amt) {
+    function collect(address[] memory assets, uint256 amount) external returns (uint256 amt) {
         require(king != address(0), "D3MAaveDaiPool/king-not-set");
 
         amt = rewardsClaimer.claimRewards(assets, amount, king);
         emit Collect(king, assets, amt);
     }
 
     function transferShares(address dst, uint256 amt) external override returns (bool) {
-        return ShareTokenLike(share).transfer(dst, amt);
+        return ShareTokenLike(adai).transfer(dst, amt);
```

Comment:
i thought so too but second guessed.  This function is (currently) only called in `exit` in the Hub and deals with `vat.slip` which should be `gems` and therefore shares, right?

Reply:
i think we are passing in amt of `assets`.  The old code used `adai.balanceOf(address(this))` which is currently our `assetBalance` function: https://github.com/makerdao/dss-direct-deposit/blob/master/src/DssDirectDepositAaveDai.sol#L496

After changing `shareBalance` to use aave's scaled balance this change works: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/1cb8e20f91b56fea64d4bd5b25cc34a9b46a3a11

Reply:
mm I think this is wrong now. TLDR:
- `transferShares()` should get shares amount as a parameter, not assets amount.
- `shareBalance()` should return units of `adai` and not underlying principal (just use `ShareTokenLike(adai).balanceOf(address(this)`, same as in `assetBalance()`). 
- The comparison we use `convertToShares()` for can be replaced in the hub with comparing `assetBalance()` of our pool address before and after `pool.deposit()` against amount.
- I think we can remove`convertToShares()` as if we do the above change it is unused.

In more details, think this latest change broke the reliance on eip4626 stadard/guidance where share is adai and asset is the underlying asset (dai). I think the confusion stems from our definition of shares balance and asset balance.
It should probably just be:
`shares` = adai / cdai
`asset`   = dai

In other words - shares balance is simple - always `balanceOf()` of the token - `adai`, `cdai`..
Asset balance should be the amount of underlying dai that the position is entitled to.
In aave's case this is exactly same as the amount of shares (`adai`) that he/she has.
In compound's case it is not and has a scalig factor.
Currently the code changed to use the user's principal (think of Maker's `art)` as share, which is a different definition, and I don't think we should use it.

Regarding the check around `pool.deposit(amount)`, for which we use `shareBalance` and `convertToShares` - 
I think we should change to use the underlying balance (i.e asset balance as I defined it above) for this.
This will remove the need for predicting the `adai `amount that we will get which I think complicated things here.

Reply:
To me shares is the non-rebasing part and asset is the rebasing underlying asset. You are right in the case of adai the balance of the token is the same as the underlying asset.

Assets should be the interest-included part and shares should always be static. To me it doesn't matter how they encoded this in the pool (such as using `balanceOf()` or `scaledBalanceOf()`) this should be exposed to the Hub as a consistent rebase / non-rebase pair.

If you want to refactor the hub to remove some unused functions such as `convertToShares()` that's fine by me.

Reply:
couple of ideas how we might resolve these.  I think we can keep the hub as "dai demoninated" as possible and let the shares be handled by the pools. which seems safer since sam you have mentioned there might be pools where shares don't exist

change `wind` to check dai balance: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/1c539a6bcdfcaf6041f92e4da23111ea6e6b0444

update exit/quit to use dai/gem denominations and the pool should convert to whatever needs to be converted when transfering shares.

Reply:
If we could drop the shares altogether that's certainly nicer. The issue we will run into with those code changes is that there is an off-by-one rounding error where the amount you are depositing into the Aave pool may be one more than the amount of adai you get. This is fine, but we either need to take this off-by-one error into account or drop the require in `wind()`.

Reply:
@hexonaut and @talbaneth how do you feel about this approach: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/e96e8d3a4c66cfea0aab458bc2630c9c81fb81f4

It handles rounding issues (I think) plus keeps us in asset-world instead of relying on share calculations

Reply:
> 

oh that is a great idea... now that we've separated things this should happen in the pool. will do that

Reply:
done: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/b1e2988c50c5a98df6f23769ec92a3b49065796a